### PR TITLE
Clean ParserCache and extract a static method in SourceLoader

### PR DESCRIPTION
**Pull request by eregon**
So if we want to extend the ParserCache it should be much easier now.

### CODE REVIEW

File: src/main/java/org/truffleruby/language/loader/SourceLoader.java [hunk 1]
```diff
@@ -210,13 +213,15 @@ public static Source loadResource(String path) throws IOException {
         }
     }
 
-    private void ensureReadable(String path, File file) {
-        if (!file.exists()) {
-            throw new RaiseException(context.getCoreExceptions().loadError("No such file or directory -- " + path, path, null));
-        }
+    private static void ensureReadable(RubyContext context, String path, File file) {
+        if (context != null) {
```
Comment (nirvdrum) [inline src/main/java/org/truffleruby/language/loader/SourceLoader.java]: When would the context be null if we're pre-initializing? And, I think we need to standardize on how we want to check for AOT without a context. We have `RubyContext.FIRST_INSTANCE` and that's used by `ParseRopeOperations.isBuildingNativeImage()`. Those might convey intent more clearly here.

File: src/main/java/org/truffleruby/language/loader/SourceLoader.java [hunk 2]
```diff
@@ -140,11 +140,15 @@ public Source load(String canonicalPath) throws IOException {
             Log.LOGGER.info("loading " + canonicalPath);
         }
 
+        return loadNoLogging(context, canonicalPath, isInternal(canonicalPath));
+    }
+
+    public static Source loadNoLogging(RubyContext context, String canonicalPath, boolean internal) throws IOException {
```
Comment (nirvdrum) [inline src/main/java/org/truffleruby/language/loader/SourceLoader.java]: Is the significance of "no logging" really that the context isn't being used?

File: src/main/java/org/truffleruby/language/loader/SourceLoader.java (hunk 1 above)
Comment (nirvdrum) [reply src/main/java/org/truffleruby/language/loader/SourceLoader.java]: I think nooping if the context is null can lead to some difficult to debug situations. Maybe throw a different Java-level exception in that case? I'd suggest assertions, but that can drastically increase native image compilation.

Comment (eregon) [reply src/main/java/org/truffleruby/language/loader/SourceLoader.java]: It's from ParserCache which doesn't have a context.
The exceptions will be thrown as some kind of IOException when reading anyway, so I think this nice early checks are not needed for the ParserCache case.

File: src/main/java/org/truffleruby/language/loader/SourceLoader.java (hunk 2 above)
Comment (eregon) [reply src/main/java/org/truffleruby/language/loader/SourceLoader.java]: This is used both with a null and non-null context.
The only part completely ignored in this method from the original is the logging, hence the name.
It's a static version of load() handling whether context is null or not.

**Pull request merged**